File: gmind/frontend/src/store/notes.ts
import { create } from 'zustand'
import { notesApi, type Note, type CreateNoteRequest } from '../api/notes'

interface NotesState {
  notes: Note[]
  loading: boolean
  error: string | null
  searchQuery: string

  fetchNotes: (query?: string) => Promise<void>
  createNote: (req: CreateNoteRequest) => Promise<Note>
  updateNote: (id: string, updates: { content?: string; tags?: string[]; pinned?: boolean }) => Promise<void>
  deleteNote: (id: string) => Promise<void>
  setSearchQuery: (q: string) => void
}

export const useNotesStore = create<NotesState>((set, get) => ({
  notes: [],
  loading: false,
  error: null,
  searchQuery: '',

  fetchNotes: async (query) => {
    set({ loading: true, error: null })
    try {
      const notes = await notesApi.list(query)
      set({ notes: notes ?? [] })
    } catch (e) {
      set({ error: String(e) })
    } finally {
      set({ loading: false })
    }
  },

  createNote: async (req) => {
    const note = await notesApi.create(req)
    set(s => ({ notes: [note, ...s.notes] }))
    return note
  },

  updateNote: async (id, updates) => {
    const updated = await notesApi.update(id, updates)
    set(s => ({
      notes: s.notes.map(n => n.id === id ? updated : n),
    }))
  },

  deleteNote: async (id) => {
    await notesApi.delete(id)
    set(s => ({ notes: s.notes.filter(n => n.id !== id) }))
  },

  setSearchQuery: (q) => {
    set({ searchQuery: q })
    get().fetchNotes(q || undefined)
  },
}))

File: gmind/backend/internal/api/notes.go
package api

import (
	"encoding/json"
	"net/http"

	"github.com/go-chi/chi/v5"
	"github.com/gmind/backend/internal/store"
)

func (h *Handler) noteStore() *store.NoteStore {
	return store.NewNoteStore(h.store.DB())
}

func (h *Handler) ListNotes(w http.ResponseWriter, r *http.Request) {
	query := r.URL.Query().Get("q")
	notes, err := h.noteStore().List(query)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(notes)
}

func (h *Handler) CreateNote(w http.ResponseWriter, r *http.Request) {
	var req store.CreateNoteRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "invalid request", http.StatusBadRequest)
		return
	}
	if req.Content == "" {
		http.Error(w, "content required", http.StatusBadRequest)
		return
	}
	note, err := h.noteStore().Create(req)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(note)
}

func (h *Handler) UpdateNote(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "noteID")
	var req store.UpdateNoteRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "invalid request", http.StatusBadRequest)
		return
	}
	note, err := h.noteStore().Update(id, req)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	if note == nil {
		http.Error(w, "not found", http.StatusNotFound)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(note)
}

func (h *Handler) DeleteNote(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "noteID")
	if err := h.noteStore().Delete(id); err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}
